qcodes/instrument_drivers/tektronix/keithley_7510.py

```diff
@@ -1,10 +1,367 @@
-from typing import cast
+import numpy as np
+from typing import cast, Optional, List, Union, Any
 
-from qcodes import VisaInstrument, InstrumentChannel
-from qcodes.utils.validators import Enum, Numbers, Ints
+from qcodes import VisaInstrument, InstrumentChannel, ParameterWithSetpoints
+from qcodes.instrument.parameter import invert_val_mapping, Parameter, \
+    DelegateParameter, MultiParameter
+from qcodes.utils.validators import Enum, Numbers, Ints, Lists, Arrays
 from qcodes.utils.helpers import create_on_off_val_mapping
 
 
+class DataArray7510(MultiParameter):
+    """
+    Data class when user selected more than one field for data output.
+    """
+    value = None
+
+    def __init__(self,
+                 names: tuple,
+                 shapes: tuple,
+                 setpoints: tuple,
+                 **kwargs):
+        self.names = names
+        super().__init__(name='data_array_7510',
+                         names=names,
+                         shapes=shapes,
+                         setpoints=setpoints,
+                         **kwargs)
+        for param_name in self.names:
+            self.__dict__.update({param_name: []})
+
+    def set_raw(self, value: tuple) -> None:
+        self.value = value
+        for i in range(len(self.names)):
+            setattr(self, self.names[i], value[i])
+
+    def get_raw(self) -> Optional[tuple]:
+        return self.value
+
+
+class GeneratedSetPoints(Parameter):
+    """
+    A parameter that generates a setpoint array from start, stop and num points
+    parameters.
+    """
+    def __init__(self,
+                 start: Parameter,
+                 stop: Parameter,
+                 n_points: Parameter,
+                 *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._start = start
+        self._stop = stop
+        self._n_points = n_points
+
+    def get_raw(self):
+        return np.linspace(self._start(), self._stop(), self._n_points())
+
+
+class ParameterWithSetpointsCustomized(ParameterWithSetpoints):
+    """
+    A Parameter class which can includes string values.
+    """
+    def validate(self, value: Any) -> None:
+        """
+        Overwrites the standard ``validate`` method to only check the the
+        parameter has consistent shape with its setpoints.
+
+        Note: the validation of the datatype is removed!
+        """
+        self.validate_consistent_shape()
```

Comment:
Why is this hack needed?

Reply:
For some of the "elements", for example, "measurement_formatted", the return values would have letters in addition of numbers: `['+00.010 V', '+00.011 V'...]`. It can not be converted to numbers, and has to be kept as list of str, which won't pass the validation check of `ParameterWithSetpoints`.

This is only for the situation when only one element is selected, because for more than one, `MultiParameter` works with list of str. Unfortunately, `MultiParameter` has to have more than one values, so for 1-elements, I have to hack the `ParameterWithSetpoints`.

Suggestions?

Reply:
I don't understand why you say that. There is no requirement that the Multiparameter must have more than one value

Reply:
You are absolutely right. Now only the `MultiParameter` is used, no need for `ParameterWithSetpoints`.